loki/transformations/data_offload/offload.py

```diff
@@ -148,28 +153,40 @@ def insert_data_offload_pragmas(self, routine, targets):
+                pragma_post = None
+                if self.directive == 'openacc':
+                    if self.present_on_device:
+                        if self.assume_deviceptr:
+                            offload_args = inargs + outargs + inoutargs
+                            if offload_args:
+                                deviceptr = f' deviceptr({", ".join(offload_args)})'
+                            else:
+                                deviceptr = ''
+                            pragma = Pragma(keyword='acc', content=f'data{deviceptr}')
                         else:
-                            deviceptr = ''
-                        pragma = Pragma(keyword='acc', content=f'data{deviceptr}')
+                            offload_args = inargs + outargs + inoutargs
+                            if offload_args:
+                                present = f' present({", ".join(offload_args)})'
+                            else:
+                                present = ''
+                            pragma = Pragma(keyword='acc', content=f'data{present}')
                     else:
-                        offload_args = inargs + outargs + inoutargs
-                        if offload_args:
-                            present = f' present({", ".join(offload_args)})'
-                        else:
-                            present = ''
-                        pragma = Pragma(keyword='acc', content=f'data{present}')
-                else:
-                    copyin = f'copyin({", ".join(inargs)})' if inargs else ''
-                    copy = f'copy({", ".join(inoutargs)})' if inoutargs else ''
-                    copyout = f'copyout({", ".join(outargs)})' if outargs else ''
-                    pragma = Pragma(keyword='acc', content=f'data {copyin} {copy} {copyout}')
-                pragma_post = Pragma(keyword='acc', content='end data')
+                        copyin = f'copyin({", ".join(inargs)})' if inargs else ''
+                        copy = f'copy({", ".join(inoutargs)})' if inoutargs else ''
+                        copyout = f'copyout({", ".join(outargs)})' if outargs else ''
+                        pragma = Pragma(keyword='acc', content=f'data {copyin} {copy} {copyout}')
+                    pragma_post = Pragma(keyword='acc', content='end data')
+                elif self.directive == 'omp-gpu':
+                    if self.present_on_device:
+                        ... # TODO: OpenMP offload if self.present_on_device
+                    else:
+                        copyin = f'map(to: {", ".join(inargs)})' if inargs else ''
+                        copy = f'map(tofrom:{", ".join(inoutargs)})' if inoutargs else ''
+                        copyout = f'map(from: {", ".join(outargs)})' if outargs else ''
+                        pragma = Pragma(keyword='omp', content=f'target data {copyin} {copy} {copyout}')
+                    pragma_post = Pragma(keyword='omp', content='end target data')
```

Comment:
I think something that would be great here would be an abstraction of the encoded concept: We inject statements to perform data allocation and movement, by 
1) specifying a list of variables that are in/out/inout/create.
2) specifying the programming model.
3) specifying the IR (PragmaRegion) to apply this to

I would create an abstract interface that takes this information and applies it to a region, with implementations for OpenACC, OpenMP, FIELD_API etc. That makes this easier to test standalone and then to re-use everywhere (offload trafo, pool allocator, global var offload, ...)

Reply:
Yes, I think we have a rudimentary version of this for FIELD-API only already, although it would probably need extending/generalising:
https://github.com/ecmwf-ifs/loki/blob/main/loki/transformations/field_api.py#L33

 I fully agree though, that having a common abstraction for this would be great.